src/Messages/Interactions/TSInteraction.h

```diff
@@ -33,7 +32,18 @@ NS_ASSUME_NONNULL_BEGIN
 + (instancetype)interactionForTimestamp:(uint64_t)timestamp
                         withTransaction:(YapDatabaseReadWriteTransaction *)transaction;
 
-- (nullable NSDate *)receiptDateForSorting;
+// NSDate has second precision, uint64_t timestamps have millisecond precision
+// so prefer timestampForSorting over dateForSorting.
+- (NSDate *)dateForSorting;
```

Comment:
Great point. 

> Now that it's fixed, do you still want to make all the changes related to the addition of the timestampForSorting?

I think so.  Why?

* We're already heavily committed to using `uint64_t` for `timestamp`.  It's nice to migrate `receivedAtTimestamp` to the same date type for consistency.
* We're carefully constructing the timestamps of the dynamic interactions to control the ordering of interactions in message view (See: `kBlockOfferOffset`, `kAddToContactsOfferOffset`, `kUnreadIndicatorOfferOffset`, etc.).  Here too it's nice to use a consistent datatype.
* I've just added a `compareForSorting` method to TSInteraction that dries up a lot of sorting logic.